STORM-1886 Extend KeyValueState iface with delete

kosii opened the pull request:
The patch also provides implementation for delete in
RedisKeyValueState and InMemoryKeyValueState.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -77,7 +81,8 @@ public RedisKeyValueState(String namespace, JedisCommandsInstanceContainer jedis
         this.keySerializer = keySerializer;
         this.valueSerializer = valueSerializer;
         this.jedisContainer = jedisContainer;
-        this.pendingPrepare = new ConcurrentHashMap<>();
```
Could we somehow handle the deletes by maintaining an entry (with some extra info) within the pendingPrepare itself and avoid the extra hash sets and synchronized ? Something along the lines of maintaining a map of
`key -> {ADD or DELETE, value}`. The values to be deleted could be written to a separate Redis Set during prepare.

During `delete`, insert a DELETE entry into pendingPrepare.
During `prepare`, snapshot pendingPrepare to pendingCommit and also save the DELETE entries to a redis set.
During `commit`, delete the DELETE entries from redis map.
During `get`, if key is in pendingPrepare, pendingCommit with DELETE or the key is not found return null.

kosii replied:
Using guava's `Optional` seems like a good idea to you?

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -104,9 +109,11 @@ private void initPendingCommit() {
             if (commands.exists(prepareNamespace)) {
                 LOG.debug("Loading previously prepared commit from {}", prepareNamespace);
                 pendingCommit = Collections.unmodifiableMap(commands.hgetAll(prepareNamespace));
+                pendingDeleteCommit = Collections.emptySet();
```
during init, shouldn't the pendingDeleteCommit also be restored from previously saved state ?

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -104,9 +109,11 @@ private void initPendingCommit() {
             } else {
                 LOG.debug("No previously prepared commits.");
                 pendingCommit = Collections.emptyMap();
+                pendingDeleteCommit = Collections.emptySet();
```
pendingDeleteCommits should probably be saved to a separate namespace say a redis set?

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -116,20 +123,30 @@ private void initPendingCommit() {
-        } else {
+        boolean found = false;
+        synchronized (this) {
+            if (pendingPrepare.containsKey(redisKey) || pendingDeletePrepare.contains(redisKey)) {
```
if pendingDeletePrepare contains the key the value should not be returned.

kosii replied:
if `pendingDeletePrepare` contains the key, then it was already removed from `pendingPrepare`, so it will return `null`

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -163,15 +190,20 @@ public void prepareCommit(long txid) {
-            if (!pendingPrepare.isEmpty()) {
+            if (!pendingPrepare.isEmpty() || !pendingDeletePrepare.isEmpty()) {
                 commands.hmset(prepareNamespace, pendingPrepare);
+                commands.hdel(prepareNamespace, pendingDeletePrepare.toArray(new String[pendingDeletePrepare.size()]));
```
Shouldn't you save the pendingDeletePrepare keys separately into a redis hashset. If a previously saved key (only in redis) is deleted, this may not work.

kosii replied:
Yes, you are right, I forget easily that the `pendingPrepare`/`pendingCommit` doesn't actually contain the snapshot, just a delta

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -184,15 +216,17 @@ public void commit(long txid) {
             commands = jedisContainer.getInstance();
-            if (!pendingCommit.isEmpty()) {
+            if (!pendingCommit.isEmpty() || !pendingDeleteCommit.isEmpty()) {
                 commands.hmset(namespace, pendingCommit);
```
Redis may throw error if pendingCommit is empty. you may want to confirm.

kosii replied:
Yes, you're right!

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -203,12 +237,14 @@ public void commit() {
             commands = jedisContainer.getInstance();
-            if (!pendingPrepare.isEmpty()) {
+            if (!pendingPrepare.isEmpty() || !pendingDeletePrepare.isEmpty()) {
                 commands.hmset(namespace, pendingPrepare);
```
Redis may throw error if pendingPrepare is empty. you may want to confirm.

kosii replied:
Yes, you're right!

arunmahadevan commented on storm-core/src/jvm/org/apache/storm/state/KeyValueState.java:
```diff
@@ -45,4 +45,11 @@
+     *
+     * @param key   the key
+     */
+    void delete(K key);
```
Could probably be changed to `V remove(K key)` along the lines of java.util.Map.

kosii replied:
It was my initial idea, but in case of the Redis implementation it needs and extra database request , as the key might not currently in one of the `pending*` structures. Does the overhead of the extra query looks negligible to you?

arunmahadevan commented:
@kosii the delete api is useful. Lets also try to see if we could handle the deletes by maintaining an entry within the pendingPrepare itself and avoid the extra hashsets and synchronized as mentioned in https://github.com/apache/storm/pull/1470/files#r66784961

kosii commented:
Thanks @arunmahadevan for the thorough review, I'm gonna address each comments asap

kosii commented:
A short list of modifications

- I went for the `Optional` class. Using your idea largely simplified the code.
- delete now returns the current value
- I didn't use a different set to store `pendingPrepare`, the deleted keys are represented with tombstones in the redis hash

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -60,22 +64,23 @@ public RedisKeyValueState(String namespace) {
 
     public RedisKeyValueState(String namespace, JedisCommandsInstanceContainer jedisContainer,
-                              Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+                              Serializer<K> keySerializer, Serializer<Optional<V>> valueSerializer) {
```
It might be better to keep the Optional internal to the implementation.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -60,22 +64,23 @@ public RedisKeyValueState(String namespace) {
         this.txidNamespace = namespace + "$txid";
         this.keySerializer = keySerializer;
         this.valueSerializer = valueSerializer;
+        this.tombstone = encode(valueSerializer.serialize(Optional.<V>absent()));
```
Can `tombstone` be a static final constant ?

kosii replied:
I'm not sure, but I don't think so, as the `internalValueSerializer` would depend on the `V` type parameter

arunmahadevan replied:
Internally absent seems to return a static final Absent<Object> cast to Optional<T>, so the type does not really matter. Its not major, but the tombstone could be a constant reference to Optional.<Object>absent().The current one is also fine.

arunmahadevan commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -161,7 +176,11 @@ public void prepareCommit(long txid) {
+                }
             }
             if (!pendingPrepare.isEmpty()) {
                 commands.hmset(prepareNamespace, pendingPrepare);
```
A potential issue (already existing) which could be fixed here or in a separate JIRA. If the pendingPrepare is updated after the hmset but before its assigned to pendingCommit, the pendingCommit will not reflect the entities in prepareNamespace. I think if the pendingPrepare is snapshotted/copied before hmset and the copy is used for both hmset and assigned to pendingCommit it would avoid this race condition.

kosii replied:
I'd fix it in a new JIRA, that way it might be released sooner

arunmahadevan replied:
Sure

arunmahadevan commented:
@kosii I have gone through the latest changes and left a few comments. The idea of using tombstone looks nice and once the comments are addressed its good to merge.

kosii commented on external/storm-redis/src/main/java/org/apache/storm/redis/state/RedisKeyValueState.java:
```diff
@@ -60,22 +64,23 @@ public RedisKeyValueState(String namespace) {
     }
 
-    public RedisKeyValueState(String namespace, JedisPoolConfig poolConfig, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+    public RedisKeyValueState(String namespace, JedisPoolConfig poolConfig, Serializer<K> keySerializer, Serializer<Optional<V>> valueSerializer) {
```
by having a `private final Serializer<Optional<V>> internalValueSerializer`?

arunmahadevan replied:
Sounds good.  If I understood correctly, the user provided serializer would be used to serialize the value to bytes which is then wrapped in an Optional<byte[]> and serialized via the DefaultStateSerializer.

HeartSaVioR commented:
@kosii Any update on this?

kosii commented:
@HeartSaVioR I addressed most of the comments in my local repo, but not pushed yet. If I have some time this week (probably Friday) I'll clean it up, and commit. Sorry for the delay

arunmahadevan commented:
@kosii can you up-merge and push your local repo changes so that we can merge this?

kosii commented:
@arunmahadevan I'm very sorry for disappearing for such a long time. I addressed all the comments, I also fixed the race condition in my last commit

arunmahadevan commented:
@kosii thanks for working on this. Will take a look one final time and merge it if it looks fine.

arunmahadevan commented:
+1

arunmahadevan commented:
@kosii thanks for the PR. Merged it to master. 

Can you also raise a PR for 1.x-branch ? I am getting some merge conflicts while trying to cherry-pick.

kosii commented:
Yes, I'll look into that!

kosii commented:
@arunmahadevan actually my patch depends on #2020, which isn't yet in 1.x-branch. how should we proceed?

arunmahadevan commented:
@kosii can you cherry pick the commits from 1606 plus your commits and raise a pr for 1.x branch?

kosii commented:
@arunmahadevan yes, check #1811 please

arunmahadevan commented:
Thanks @kosii Merged to master and 1.x-branch. You can close the PR.
